Add correction parameter to std/var

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55681 std/var: Deprecate overloads with "unbiased" argument
* #55680 std/var: Require correction or unbiased parameter be given
* #55679 std/var: Deprecate default correction value
* #57281 std/var: Return real results for complex input
* #56398 Give std/var correction overloads proper defaults
* **#50903 Add correction parameter to std/var**

First part of #50010. Also fixes #51127.

Added overloads for torch.{std, var, std_mean, var_mean} with a correction argument specifying the difference between the sample size and number of degrees of freedom. e.g `correction=1` is equivalent to Bessel's correction, which can also be gained using the `unbiased=True` overload.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

peterbell10 committed:
Add correction parameter to std/var

[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

mruberry commented:
Hey @peterbell10! Thanks for updating this. @heitorschueroff and I will take a look early this week!

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 commented:
Added onnx test with correction parameter and rebased on latest.

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1205,8 +1206,11 @@ Tensor argmin(const Tensor& self, c10::optional<int64_t> dim, bool keepdims) {
+    const char* fname, Tensor& result, const Tensor& self,
+    c10::optional<IntArrayRef> dim, c10::optional<int64_t> correction_opt,
+    bool keepdim, bool take_sqrt) {
+  TORCH_CHECK(self.device().is_cpu() || self.device().is_cuda(),
```
While we're here, can we tweak these error messages:

"std and var only supports tensors a CPU or CUDA device, but got: ..."

and the error message for the type check needs a tweak, too:

"std and var only support floating point and complex dtypes"

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1217,35 +1221,59 @@ static Tensor& std_var_out(Tensor& result, const Tensor& self, IntArrayRef dim,
-    } else {
-      std_var_stub(iter.device_type(), iter, unbiased, false);
-    }
+    std_var_out(
```
Would you add a comment before this block explaining the std + var computation for complex so readers know what to expect

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1217,35 +1221,59 @@ static Tensor& std_var_out(Tensor& result, const Tensor& self, IntArrayRef dim,
+    return result;
+  }
+
+  const auto correction = correction_opt.value_or(1);
```
Add a comment here that this is the computation for the floating point case

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1217,35 +1221,59 @@ static Tensor& std_var_out(Tensor& result, const Tensor& self, IntArrayRef dim,
+    bool keepdim, bool take_sqrt) {
   AT_ASSERT(result1.defined() && result2.defined());
   TORCH_CHECK(self.device().is_cpu() || self.is_cuda(),
               fname, " only supports CPU AND CUDA device type, got: ", self.device().type());
```
Similar tweak to error message here as above

mruberry commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -193,6 +193,8 @@ static TensorIterator make_reduction(
       toString(out_dtype),
       ".");
+  // dim={} performs an all-reduce, same as dim=None
+  IntArrayRef dim = dim_opt.value_or(IntArrayRef{});
```
Cool update

mruberry commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -157,13 +157,17 @@ static void mean_kernel_impl(TensorIterator& iter) {
 }
 
-static void std_var_kernel_impl(TensorIterator &iter, bool unbiased, bool take_sqrt) {
+static void std_var_kernel_impl(TensorIterator& iter, int64_t correction, bool take_sqrt) {
```
Thank you for putting the ampersand with the type ;)

mruberry commented on aten/src/ATen/native/cuda/ReduceMomentKernel.cu:
```diff
@@ -2,35 +2,33 @@
+  using limits = std::numeric_limits<int32_t>;
+  TORCH_CHECK(
+      correction < limits::max() && correction > limits::min(),
+      "Correction must fit in 32-bit on CUDA, but got ",
```
Cool check. Suggested language tweak"

"The correction argument for std and var computation on CUDA must fit within a 32-bit integer, but got ..."

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6276,14 +6303,6 @@
     CPU: _cumprod_out_cpu
     CUDA: _cumprod_out_cuda
 
-- func: _var(Tensor self, bool unbiased=True) -> Tensor
```
Thank you for removing these from the public API

mruberry commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -300,18 +301,20 @@ void THTensor_(renorm)(THTensor *res, THTensor *src, scalar_t value, int dimensi
-{
+accreal THTensor_(std_var_all)(THTensor* tensor, int64_t correction, bool take_sqrt) {
+  const auto denom = THTensor_(nElement)(tensor) - correction;
+  if (denom <= 0) {
```
Nice refactor

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8186,45 +8186,115 @@ def merge_dicts(*dicts):
-std(input, unbiased=True) -> Tensor
+std(input, dim=None, *, correction=1, keepdim=False, out=None) -> Tensor
+
+Returns the standard-deviation over the dimensions specified by :attr:`dim`.
```
"Returns" -> "Computes"

mruberry replied:
"standard-deviation" -> "standard deviation"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8186,45 +8186,115 @@ def merge_dicts(*dicts):
+    {dim}
+
+Keyword args:
+    correction (int): difference between the sample size and sample degrees of freedom.
```
This is an excellent and clear description

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8186,45 +8186,115 @@ def merge_dicts(*dicts):
-Returns the standard-deviation of each row of the :attr:`input` tensor in the
-dimension :attr:`dim`. If :attr:`dim` is a list of dimensions,
-reduce over all of them.
+Returns the standard-deviation and mean over the dimensions specified by
```
"standard-deviation" -> "standard deviation"

mruberry commented:
Hey @peterbell10!

This is an awesome PR. @heitorschueroff and I had a chance to go through most of it and, as usual, it looks really good. We made a few inline suggestions/comments. @heitorschueroff is going to follow-up and continue reviewing the autograd, ONNX, testing and jit stuff.

I really appreciated the way you explained "correction" in the docs, btw.

heitorschueroff reviewed (commented):
@peterbell10 Mike and I reviewed the PR together. It's looking good for the most part. I still need to look at tests, docs and some other changes which I will be reviewing in a follow up soon. The test failures are relevant.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -300,18 +301,20 @@ void THTensor_(renorm)(THTensor *res, THTensor *src, scalar_t value, int dimensi
-{
+accreal THTensor_(std_var_all)(THTensor* tensor, int64_t correction, bool take_sqrt) {
+  const auto denom = THTensor_(nElement)(tensor) - correction;
+  if (denom <= 0) {
```
Actually, no this should be dividing by zero.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

heitorschueroff commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1264,157 +1292,162 @@ static std::tuple<Tensor&,Tensor&> std_var_mean_out(const char* fname, Tensor &r
+std::tuple<Tensor, Tensor> var_mean(
+    const Tensor& self, IntArrayRef dim, bool unbiased, bool keepdim) {
+  return at::var_mean(self, /*dim=*/c10::optional<IntArrayRef>(dim),
+                      /*correction=*/int64_t{unbiased ? 1 : 0}, keepdim);
```
Oh I didn't notice the curly braces. You're right, thanks for the reference.

heitorschueroff commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3460,30 +3460,38 @@
+- func: std.dim(Tensor self, int[1] dim, bool unbiased, bool keepdim=False) -> Tensor
+  variants: function, method
+
+- func: std.correction(Tensor self, int[1]? dim=None, *, int? correction=None, bool keepdim=False) -> Tensor
```
Forward compatibility is when a newer writer continues to work with an older reader. In this case, by making unbiased a required parameter, calls to std would go into the new signature and be written (when saving models) as such. But we have internal servers that load models which won't be updated for about two weeks and so when they try to load this model they wouldn't recognize the new singature.

peterbell10 replied:
@heitorschueroff I've removed the breaking change to jit signatures from this PR.

heitorschueroff replied:
Thank you, I'm taking a second look at the PR now.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

heitorschueroff commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -300,18 +301,17 @@ void THTensor_(renorm)(THTensor *res, THTensor *src, scalar_t value, int dimensi
-accreal THTensor_(std_all)(THTensor *tensor, bool unbiased)
-{
-  return sqrt(THTensor_(var_all)(tensor, unbiased));
+  sum /= std::max(int64_t{0}, THTensor_(nElement)(tensor) - correction);
```
Nice refactor combining these methods.

heitorschueroff commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -83,6 +83,8 @@
     ("aten::empty_meta", datetime.date(2021, 4, 1)),
     ("aten::div", datetime.date(2021, 4, 28)),
     ("aten::divide", datetime.date(2021, 4, 28)),
+    ("aten::_var", datetime.date(2021, 4, 28)),
```
We shouldn't have backward compatibility issues since we are no longer modifying the existing signatures.

peterbell10 replied:
Doesn't the CI check fail when you remove a signature?

heitorschueroff commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -907,56 +907,60 @@ Tensor evenly_distribute_backward(Tensor grad, const Tensor & input, const Tenso
-    Tensor mean_grad = mean_backward(grads[1], self.sizes(), dim, keepdim);
-    grad = grads[0].defined() ? grad + mean_grad : mean_grad;
+static Tensor mean_backward(
+    const Tensor& grad, const Tensor& self, c10::optional<IntArrayRef> dim, bool keepdim) {
```
We should pass self.sizes() as parameter to be consistent with the other `mean_backward` definitions.

heitorschueroff commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2949,10 +2953,10 @@ infinitely_differentiable_native_layer_norm_backward(
-            false,
-            true,
-            false);
+            /*dim=*/IntArrayRef{1},
```
Nice comments

heitorschueroff commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -342,7 +378,8 @@ const std::vector<std::string> functions = {
                 return grad_self, grad_mat2
             return torch.bmm(self, mat2), backward
-
+    )",
+    R"(
```
What are these two lines for? I'm confused.

peterbell10 replied:
There's a max string literal size on MSVC. So, I'm splitting the string to get below the size limit.

heitorschueroff reviewed (commented):
@peterbell10 I finished another review pass. The PR looks very good for the most part. Here are some highlights

- I'm not sure we have completely fixed the forward compatibility, @mruberry can you help verify the changes with the `cpp_no_default_args` entry in native_functions.yaml
- I left some notes on the doc mostly to consider if we want to hide the new signatures until after forward compatibility period is over and to combine overrides into one signature
- XLA failure is real, cc @JackCaoG
```
Apr 15 20:13:47 torch_xla/csrc/aten_xla_type_default.cpp:4322:22: error: no matching function for call to 'std'
Apr 15 20:13:47   auto std_out_tmp = AtenXlaType::std(self, dim, correction, keepdim);
```
- Doc build failure is also real, I think it's due to multiple .. function entries in _tensor_docs.py.
```
Apr 14 00:49:10 /var/lib/jenkins/workspace/docs/source/generated/torch.Tensor.std.rst:11: WARNING: duplicate object description of torch.std, other instance in generated/torch.std, use :noindex: for one of them
Apr 14 00:49:10 /var/lib/jenkins/workspace/docs/source/generated/torch.Tensor.var.rst:11: WARNING: duplicate object description of torch.var, other instance in generated/torch.var, use :noindex: for one of them
```

Once these issues are addressed I think the PR can be landed. 
Let me know if you have any questions and ping me when I should take another look.

Looking forward to the update and great work so far in this very challenging PR.

peterbell10 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3473,36 +3473,57 @@
     CPU, CUDA: square_out
 
 - func: std(Tensor self, bool unbiased=True) -> Tensor
+  cpp_no_default_args: ['unbiased']
```
If the problem is loading old exported models with new versions of pytorch, then isn't this fine? I don't think it's possible to introduce the new overload without the possibility of it being called, since the python API doesn't have a priority order for overloads.

heitorschueroff replied:
The problem with forward compatibility is actually loading newly exported models with old versions of pytorch. Our internal model loading servers lag behind for about 2 weeks. It's true that there's no priority order for overloads. The workaround is to make the parameters in the new signature required (no default values) for now until the internal model loading servers are updated to recognize the new signatures at which point we can remove add default values again. See the changes to native_functions in this PR https://github.com/pytorch/pytorch/pull/49267/files which faced the same problems with forward compatibility.

I know this is inconvenient and confusing and I appreciate your collaboration in this PR.

peterbell10 commented on torch/_tensor_docs.py:
```diff
@@ -3531,7 +3531,15 @@ def callable(a, b) -> number
 add_docstr_all('std',
                r"""
-std(dim=None, unbiased=True, keepdim=False) -> Tensor
+std(dim=None, *, correction=1, keepdim=False) -> Tensor
```
I don't understand. What's the point of adding it without documentation?

peterbell10 commented on torch/_torch_docs.py:
```diff
@@ -8097,105 +8097,161 @@ def merge_dicts(*dicts):
     {out}
 
+
+.. function:: std(input, unbiased) -> Tensor
```
No, `std(None, unbiased=True)` doesn't work. There is no such signature.

heitorschueroff replied:
You're right, this is a limitation of the existing signatures but that's an issue for another PR.

JackCaoG commented:
I will work on a fix on the pt/xla side

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

peterbell10 commented:
@heitorschueroff PTAL. Removed default correction value, removed correction from the docs and moved them into another PR higher up the stack.

heitorschueroff commented:
> @heitorschueroff PTAL. Removed default correction value, removed correction from the docs and moved them into another PR higher up the stack.

Thank's for the updates @peterbell10 I'll be taking a look at it later today. I'll also take a look at the other PRs on the stack later.

heitorschueroff reviewed (approved):
LGTM! @mruberry would you like to take another look at this?

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3475,28 +3475,35 @@
 - func: std_mean.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
   variants: function
+
+- func: std_mean.correction(Tensor self, int[1]? dim, *, int? correction, bool keepdim=False) -> (Tensor, Tensor)
```
These signatures look good, but I think they need to require correction `int correction` so they're only matched when `correction` is actually specified by the user

peterbell10 replied:
`int?` without a default means the user needs to explicitly pass `int` or `None`. Either way, the user needs to specify it.

peterbell10 replied:
Also, if we change to optional later when the default gets added then that requires a breaking change to the JIT schema so the "forward compatibility" issues @heitorschueroff mentions wouldn't be resolved.

mruberry commented on test/test_reductions.py:
```diff
@@ -2117,6 +2115,122 @@ def test_std_vs_numpy(self, device, dtype):
+                # dim
+                (None, 0, 1),
+                # correction
+                (None, 0, 10, 30),
```
Would you add a negative value here, too?

mruberry commented on test/test_reductions.py:
```diff
@@ -2117,6 +2115,122 @@ def test_std_vs_numpy(self, device, dtype):
+            )
+        ]
+
+        tensor = torch.rand(_size, device=device, dtype=dtype)
```
use `make_tensor()`

mruberry commented on test/test_reductions.py:
```diff
@@ -2117,6 +2115,122 @@ def test_std_vs_numpy(self, device, dtype):
+                # dim
+                (None, 0, 1),
+                # correction
+                (None, 0, 10, 30),
```
Analogous comments here

mruberry commented on tools/autograd/derivatives.yaml:
```diff
@@ -1088,11 +1088,11 @@
 
-- name: std(Tensor self, bool unbiased=True) -> Tensor
-  self: std_backward(result, grad, self, unbiased)
+- name: std.correction(Tensor self, int[1]? dim, *, int? correction, bool keepdim=False) -> Tensor
```
I'm curious if `int? correction` should be the thing here?

Are std and var going to be implemented so they call into std(..., correction=...) and var(..., corection-...)? If so, should correction be optional here?

peterbell10 replied:
> Are std and var going to be implemented so they call into std(..., correction=...) and var(..., correction-...)?

Yes, the `unbiased` overloads both convert to numerical correction and call the correction overload.

>If so, should correction be optional here?

It has to exactly match the JIT schema in native_functions.yaml

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8099,105 +8099,81 @@ def merge_dicts(*dicts):
 add_docstr(torch.std_mean,
            r"""
-std_mean(input, unbiased=True) -> (Tensor, Tensor)
+std_mean(input, dim, unbiased, keepdim=False, *, out=None) -> (Tensor, Tensor)
```
So the idea is that we're going to hide the correction kwarg in the documentation for the next couple weeks?

peterbell10 replied:
Yes, until the "forward compatibility" period is over.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -9351,104 +9327,80 @@ def merge_dicts(*dicts):
 
-If :attr:`unbiased` is ``False``, then the variance will be calculated via the
-biased estimator. Otherwise, Bessel's correction will be used.
+If :attr:`unbiased` is ``True``, Bessel's correction will be used.
```
Really nice language improvements

mruberry reviewed (commented):
Hey @peterbell10! I made a few comments, mostly asking questions. Let me know if it'd be easier to VC briefly to discuss some of these issues -- maybe I'm missing something?

Thanks for working on a fix on the XLA side, @JackCaoG! Please keep us informed. Do you have a rough timeline for how long it will take? Like, 1 week, 2 weeks, a month...?

peterbell10 commented:
XLA fix is pytorch/xla#2891, and CI is green so I think it's ready now.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.


[ghstack-poisoned]

mruberry reviewed (approved):
Awesome!

I'll coordinate landing this with @JackCaoG

mruberry commented:
Looks like a complex test is failing internally:

```
test_std_correction_vs_numpy_cpu_complex64 (test_reductions.TestReductionsCPU) (architecture: x86_64, buildmode: opt, buildsystem: buck, compiler: clang, sanitizer: none)


AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part nan and inf gives a difference of nan, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only inf!

torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
test_reductions.py", line 2174, in test_std_correction_vs_numpy
    self.assertEqual(torch_res, numpy_res, exact_dtype=False)
```

Maybe the test is sensitive to different random values, a different version of NumPy, or a different compiler? We might need to make the test "easier" or simplify it. The complex double version of the test is failing with the same error.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

peterbell10 commented:
> Maybe the test is sensitive to different random values, a different version of NumPy, or a different compiler? We might need to make the test "easier" or simplify it. The complex double version of the test is failing with the same error.

@mruberry the test is sensitive to machine precision, but only in the event the variance is zero (since exactly zero gives nan, while finite but small gives inf). However, zero variance seems very unlikely on random data so that's still a bit surprising and might be worth looking into.  	

For now, I've changed the test to treat nan and inf as equal.

mruberry commented:
Unfortunately the same tests are failing:

```
AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary part nan and 0.0 gives a difference of nan, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1e-05!
```

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

peterbell10 commented:
@mruberry I've loosened the restrictions on the imaginary component this time. Allowing to to be a non-finite value if the real component also isn't finite. If that doesn't work, I really need to know what arguments to `std` are failing. Including the tensor shape and contents.

mruberry commented:
> @mruberry I've loosened the restrictions on the imaginary component this time. Allowing to to be a non-finite value if the real component also isn't finite. If that doesn't work, I really need to know what arguments to `std` are failing. Including the tensor shape and contents.

Sounds good; let's see what the internal tests say.

mruberry commented:
Good news! Internal tests are now passing. I'll get the land process started tomorrow morning and ping @JackCaoG with updates

mruberry commented:
Darn, looks like the PR CI failures for fx are real:

test_normalize_operator_exhaustive_std_cpu_float32
test_normalize_operator_exhaustive_var_cpu_float32

Probably just a tweak to that test is needed?

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

peterbell10 commented:
@mruberry since the fx PR got reverted, I've removed it from this stack and instead just added a skip in the test for `std` and `var`. That should unblock this PR.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

mruberry commented:
Tests look good. @Chillee's fix may also have resolved the fx issue but we don't need to worry about unskipping it for this PR. Internal tests look. I'll check-in with @JackCaoG during business hours to validate we're OK to land this, and we'll try to merge it today.

mruberry commented:
We'll have to try again on Monday, sorry @peterbell10. Some landing issues let this accumulate another merge conflict. Let's rebase it Monday morning and I'll start the land process while no one's committing.

peterbell10 committed:
Update on "Add correction parameter to std/var"


First part of #50010. Also fixes #51127.

Differential Revision: [D27911345](https://our.internmc.facebook.com/intern/diff/D27911345)

[ghstack-poisoned]

peterbell10 commented:
@mruberry I've rebased on viable/strict and fixed merge conflicts.

mruberry commented:
Internal tests look good; unfortunately I lost power this morning, but I'll try to coordinate this with @JackCaoG tomorrow morning
